app/lib/three_scale/swagger/specification.rb

```diff
@@ -133,27 +150,24 @@ def initialize(body)
         @version = init_version
       end
 
-      delegate :base_path, :validate!, :swagger?, to: :@version
-
+      delegate :base_path, :servers, :validate!, :swagger?, to: :@version
 
       # Check if this specification is swagger thus it can be displayed in swagger-ui
       def swagger_1_2?
-        @doc.fetch("swaggerVersion", 0).to_f >= 1.2
+        doc_version.to_f >= 1.2
       end
 
       def swagger_2_0?
-        @doc.fetch("swagger", 0).to_f >= 2.0
+        doc_version.to_f >= 2.0
+      end
+
+      def openapi_3_0?
+        doc_version.to_f >= 3.0
       end
 
       # Falls back to "1.0" if version can"t be determined
       def swagger_version
-        if swagger_1_2?
-          @doc["swaggerVersion"]
-        elsif swagger_2_0?
-          @doc["swagger"]
-        else
-          "1.0"
-        end
+        (swagger_1_2? || swagger_2_0? || openapi_3_0?) ? doc_version.scan(/\A(\d+\.\d+)(\..+)?\Z/).flatten.first : '1.0'
```

Comment:
The 'else' scenario is lacking a test.

However, I did it locally to check what would happen with a change of code here. I copied and pasted the file like `cp test/fixtures/swagger/pet-store-3.0.json test/fixtures/swagger/pet-store-undefined.json` and I changed the line of `"openapi": "3.0",` for `"openapi": "-",`.
Then I wrote this quick test:
```ruby
test 'pet store default version' do
    doc = load_fixture 'pet-store-undefined'
    specification = ThreeScale::Swagger::Specification.new doc

    assert_equal '1.0', specification.swagger_version
  end
```

And with that I see, that this code works:
```suggestion
       doc_version.scan(/\A(\d+\.\d+)(\..+)?\Z/).flatten.first || '1.0'
```

Reply:
Try with swagger version "1.1" and you'll see.

Reply:
ooh, I am following you now :smile: In that case, I will think in a nicer solution :thinking:

Reply:
For me even this is less ugly:
```ruby
def swagger_version
  doc_version.to_f >= 1.2 ? doc_version.scan(/\A(\d+\.\d+)(\..+)?\Z/).flatten.first : '1.0'
end
```

Which is exactly the same but doing the `doc_version.to_f >= 1.2` only once... because mathematically does not makes sense to check that a number is bigger than 1.2 OR bigger than 2 OR bigger than 3 ... because if it is bigger than 2 o 3, it is also bigger than 1.2 :confused: And it is also ugly doing `swagger_1_2? ? doc_version.scan(/\A(\d+\.\d+)(\..+)?\Z/).flatten.first : '1.0'` because it is not that understandable for the meaning.

Another possible solution but CHANGES behaviour from this (so I dunno if it suits our needs) would be:
```ruby
def swagger_version
  swagger_version = doc_version.to_f.round(1)
  swagger_version = 1.0 unless swagger_version >= 1.2
  swagger_version.to_s
end
```

However, independently of the code, I still feel that we need a test for this scenario that you mentioned of '1.1' to '1.0'

Reply:
We do have a test for the "v1.1 not-supported" case: https://github.com/3scale/porta/blob/f5dcf0bf954202a5412523035002daa38cec6eed/test/unit/api_docs/service_test.rb#L370

This check in `ThreeScale::Swagger::Specification#swagger_version` is not strictly the same thing as "we don't support swagger 1.1". The check should be read as "give me the swagger version in the spec, as long as the spec if 1.2 or newer".

I refactored following your suggestion to make this more clear. Thanks, @Martouta !